csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs

```diff
@@ -0,0 +1,195 @@
+
+                if (footerLength <= 0) throw new InvalidDataException(
+                    $"Footer length has invalid size <{footerLength}>");
+            });
+
+            await Buffers.RentReturnAsync(footerLength, async (buffer) =>
+            {
+                _footerStartPostion = (int)BaseStream.Length - footerLength - ArrowFileConstants.Magic.Length - 4;
+
+                BaseStream.Position = _footerStartPostion;
+
+                bytesRead = await BaseStream.ReadAsync(buffer, 0, footerLength);
+
+                if (bytesRead != footerLength)
+                {
+                    throw new InvalidDataException(
+                        $"Failed to read footer. Read <{bytesRead}> bytes, expected <{footerLength}>.");
+                }
+
+                // Deserialize the footer from the footer flatbuffer
+
+                _footer = new ArrowFooter(Flatbuf.Footer.GetRootAsFooter(new ByteBuffer(buffer)));
+
+                Schema = _footer.Schema;
+            });
+        }
+
+        public async Task<RecordBatch> ReadRecordBatchAsync(int index, CancellationToken cancellationToken)
+        {
+            await ReadSchemaAsync();
+
+            if (index >= _footer.RecordBatchCount)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index));
+            }
+
+            var block = _footer.GetRecordBatchBlock(index);
+
+            BaseStream.Position = block.Offset;
+
+            return await ReadRecordBatchAsync(cancellationToken);
+        }
+
+        public override async Task<RecordBatch> ReadNextRecordBatchAsync(CancellationToken cancellationToken)
+        {
+            await ReadSchemaAsync();
+
+            if (_recordBatchIndex >= _footer.RecordBatchCount)
+            {
+                return null;
+            }
+
+            var result = await ReadRecordBatchAsync(_recordBatchIndex, cancellationToken);
+            _recordBatchIndex++;
+
+            return result;
+        }
+
+        /// <summary>
+        /// Check if file format is valid. If it's valid don't run the validation again.
+        /// </summary>
+        private async Task ValidateFileAsync()
+        {
+            if (IsFileValid)
+            {
+                return;
+            }
+
+            await ValidateMagicAsync();
+
+            IsFileValid = true;
+        }
+
+        private async Task ValidateMagicAsync()
+        {
+            var startingPosition = BaseStream.Position;
+            var magicLength = ArrowFileConstants.Magic.Length;
+
+            try
+            {
+                await Buffers.RentReturnAsync(magicLength, async (buffer) =>
+                {
+                    // Seek to the beginning of the stream
+
+                    BaseStream.Position = 0;
+
+                    // Read beginning of stream
+
+                    await BaseStream.ReadAsync(buffer, 0, magicLength);
+
+                    if (!ArrowFileConstants.Magic.SequenceEqual(buffer.Take(magicLength)))
```

Comment:
What is `Take` here?  Is that LINQ's `Take`?  If so, that's fairly expensive way to get a span to compare against, both in time and allocation.

Reply:
This was our fault :)

The synchronization code needs a lot of love. The better solution here would probably be to slice a span of the magic / buffer and compare.

Reply:
Agreed. Again this can be cleaned up with [ARROW-4503: ArrowStreamReader allocates and copies data excessively](https://issues.apache.org/jira/browse/ARROW-4503). This was just refactoring the existing code to allow for `ReadOnlyMemory<byte>` inputs and zero copies.